Examples/Docs/Sources/AsyncExamples/main.swift

```diff
@@ -0,0 +1,147 @@
+    let result1 = items.updateOne(
+        filter: ["sku": "111", "end": .null],
+        update: ["$set": ["end": .datetime(currentDate)]],
+        session: s1
+    ).flatMap { _ in
+        items.insertOne(["sku": "nuts-111", "name": "Pecans", "start": .datetime(currentDate)], session: s1)
+    }
+    // End Causal Consistency Example 1
+
+    let client2 = try MongoClient(using: elg)
+
+    // Start Causal Consistency Example 2
+    let options = ClientSessionOptions(causalConsistency: true)
+    let result2: EventLoopFuture<Void> = client2.withSession(options: options) { s2 in
+        // The cluster and operation times are guaranteed to be non-nil since we already used s1 for operations above.
+        s2.advanceClusterTime(to: s1.clusterTime!)
+        s2.advanceOperationTime(to: s1.operationTime!)
+
+        dbOptions.readPreference = ReadPreference(.secondary)
+        let items2 = client2.db("test", options: dbOptions).collection("items")
+
+        return items2.find(["end": .null], session: s2).flatMap { cursor in
+            cursor.forEach { item in
+                print(item)
+            }
+        }
+    }
+    // End Causal Consistency Example 2
+}
+
+/// Examples used for the MongoDB documentation on Change Streams.
+/// - SeeAlso: https://docs.mongodb.com/manual/changeStreams/
+private func changeStreams() throws {
+    let elg = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+    let client = try MongoClient(using: elg)
+    let db = client.db("example")
+
+    // The following examples assume that you have connected to a MongoDB replica set and have
+    // accessed a database that contains an inventory collection.
+
+    do {
+        // Start Changestream Example 1
+        let inventory = db.collection("inventory")
+        let futureCursor = inventory.watch()
+
+        // Option 1: retrieve next document via next()
+        let next = futureCursor.flatMap { cursor in
+            cursor.next()
+        }
+
+        // Option 2: register a callback to execute for each document
+        let result = futureCursor.flatMap { cursor in
+            cursor.forEach { event in
+                // process event
+                print(event)
+            }
+        }
+        // End Changestream Example 1
+    }
+
+    do {
+        // Start Changestream Example 2
+        let inventory = db.collection("inventory")
+        let futureCursor = inventory.watch(options: ChangeStreamOptions(fullDocument: .updateLookup))
+
+        // Option 1: use next() to iterate
+        let next = futureCursor.flatMap { cursor in
+            cursor.next()
+        }
+
+        // Option 2: register a callback to execute for each document
+        let result = futureCursor.flatMap { cursor in
+            cursor.forEach { event in
+                // process event
+                print(event)
+            }
+        }
+        // End Changestream Example 2
+    }
+
+    do {
```

Comment:
This example is particularly confusing due to the all the different top level futures. Here's my attempt at simplifying it a little:
```swift
inventory.watch().flatMap { changeStream in
    changeStream.next().map { _ in
        changeStream.resumeToken
    }.always { _ in
        _ = changeStream.kill()
    }
}.flatMap { resumeToken in
    inventory.watch(options: ChangeStreamOptions(resumeAfter: resumeToken)).flatMap { newStream in
        newStream.forEach { change in
            print(change)
         }
    }
}
```
What do you think?

As a side note: I think we should make `kill` be `@discardableResult`.

Reply:
that looks better than using `and`, I updated to something very similar to this.

to your aside: opened https://jira.mongodb.org/browse/SWIFT-730